File: apps/ml/ui/src/types/dataset.ts
export type FieldType = 'image' | 'video' | 'media' | 'file' | 'text' | 'number' | 'select'
export type CaptureMode = 'camera_only' | 'upload_only' | 'both'
export type DescriptionMode = 'none' | 'free_text' | 'preset'
export type DatasetStatus = 'draft' | 'active' | 'closed'

export interface DatasetField {
  id: string
  label: string
  instruction: string
  type: FieldType
  capture_mode: CaptureMode
  required: boolean
  description_mode: DescriptionMode
  description_presets: string[]
  description_required: boolean
  order: number
  repeatable: boolean
  max_repeats: number  // 0 = unlimited
  // Model validation
  validation_model: string | null
  validation_labels: string[]
  validation_message: string
}

export type DatasetVisibility = 'private' | 'public'
export type ReferenceType = 'clone' | 'reference' | null

export interface UrlDatasetInfo {
  id: string
  source_url: string
  status: 'pending' | 'fetching' | 'ready' | 'error'
  item_count: number | null
  size_bytes: number | null
  last_fetched_at: string | null
  next_fetch_at: string | null
  fetch_error: string | null
  refresh_interval_hours: number
  content_type: string
}

export interface DatasetProfile {
  id: string
  org_id: string
  name: string
  slug: string | null
  description: string
  category: string
  fields: DatasetField[]
  status: DatasetStatus
  // Visibility & sharing
  visibility: DatasetVisibility
  source_dataset_id: string | null
  reference_type: ReferenceType
  entry_count_cache: number
  discoverable: boolean
  contributor_allowlist: string[]
  points_enabled: boolean
  points_per_entry: number
  points_redemption_info: string
  // Location tracking
  require_location: boolean
  location_purpose: string
  // Consent
  require_consent: boolean
  consent_template_id: string | null
  consent_type: 'individual' | 'group'
  created_by: string
  created_at: string
  updated_at: string
  collectors?: DatasetCollector[]
  url_dataset?: UrlDatasetInfo
}

export interface DatasetCollector {
  id: string
  dataset_id: string
  email: string
  phone: string | null
  name: string
  token: string
  status: 'pending' | 'active' | 'completed'
  invited_at: string
  last_active_at: string | null
  entry_count: number
  points_earned: number
}

export type QualityIssue = 'blurry' | 'dark' | 'overexposed' | 'low_res'

export interface EntryLocation {
  lat: number | null
  lng: number | null
  accuracy: number | null
  source: 'gps' | 'ip'
  ip_address: string | null
  country: string | null
  country_name: string | null
  city: string | null
  timezone: string | null
  isp: string | null
}

export interface DatasetEntry {
  id: string
  dataset_id: string
  collector_id: string
  field_id: string
  file_key: string | null
  file_url: string | null
  file_mime: string | null
  file_size_bytes: number | null
  text_value: string | null
  description: string | null
  points_awarded: number
  location: EntryLocation | null
  captured_at: string
  review_status: 'pending' | 'approved' | 'rejected'
  review_note: string | null
  // quality metrics
  blur_score: number | null
  brightness: number | null
  quality_score: number | null
  quality_issues: QualityIssue[]
  phash: string | null
  file_hash: string | null
  archived: boolean
  consent_record_id?: string | null
}

export interface SimilarDatasetEntry extends DatasetEntry {
  similarity_distance: number
  similarity_pct: number
}

export interface DatasetEntryListResponse {
  items: DatasetEntry[]
  total: number
  page: number
  page_size: number
  total_pages: number
  archived_count: number
}

// Collect page types (public)
export interface CollectFormDefinition {
  dataset: {
    id: string
    name: string
    description: string
    category: string
    fields: DatasetField[]
    points_enabled: boolean
    points_per_entry: number
    points_redemption_info: string
    require_location: boolean
    location_purpose: string
    require_consent: boolean
    consent_type: 'individual' | 'group'
    consent_template_id: string | null
  }
  collector: {
    id: string
    name: string
    email: string
    entry_count: number
    points_earned: number
  }
}

// Dataset overview / analytics
export interface DatasetOverview {
  dataset_id: string
  name: string
  status: string
  require_location: boolean
  summary: {
    total_entries: number
    total_collectors: number
    total_points_awarded: number
    active_collectors: number
  }
  location: {
    gps_count: number
    ip_count: number
    no_location: number
    gps_pct: number
    countries: { code: string; count: number }[]
    cities: { name: string; count: number }[]
    gps_points: { lat: number; lng: number }[]
  }
  daily_trend: { date: string; count: number }[]
  top_collectors: { name: string; email: string; entries: number; points: number }[]
  field_breakdown: { field_id: string; label: string; count: number }[]
  quality: {
    scored: number
    no_score: number
    good: number
    poor: number
    avg_score: number | null
    good_pct: number
    issues: { label: string; count: number }[]
  }
}

export interface DatasetCreatePayload {
  name: string
  slug?: string
  description: string
  category: string
  fields: Omit<DatasetField, 'id'>[]
  visibility?: DatasetVisibility
  discoverable?: boolean
  contributor_allowlist?: string[]
  points_enabled: boolean
  points_per_entry: number
  points_redemption_info: string
  require_location?: boolean
  location_purpose?: string
  require_consent?: boolean
  consent_template_id?: string
  consent_type?: string
}

File: apps/ml/app/models/dataset.py
"""Dataset collection models — profile, collectors, entries."""
import uuid
import re
import secrets
from typing import Optional, List
from datetime import datetime
from beanie import Document
from pydantic import BaseModel, Field
from app.utils.datetime import utc_now


def slugify(text: str) -> str:
    """Convert text to a URL-safe slug: lowercase, spaces/underscores→hyphens, strip specials."""
    s = text.lower().strip()
    s = re.sub(r"[^\w\s-]", "", s)
    s = re.sub(r"[\s_]+", "-", s)
    s = re.sub(r"-+", "-", s).strip("-")
    return s or "dataset"


class DatasetField(BaseModel):
    """A single capture slot in the dataset (e.g. 'Front view of cow')."""
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    label: str                                    # shown to collector
    instruction: str = ""                         # guidance text / photo tips
    type: str = "image"                           # image | video | media | file | text | number
                                                  # media = accept both image + video
    # image / file settings
    capture_mode: str = "both"                    # camera_only | upload_only | both
    required: bool = True
    # description settings
    description_mode: str = "none"               # none | free_text | preset
    description_presets: List[str] = []           # e.g. ["Healthy","Sick","Injured"]
    description_required: bool = False
    order: int = 0
    # Repeat settings
    repeatable: bool = False                      # can be submitted multiple times
    max_repeats: int = 0                          # 0 = unlimited; N = cap at N submissions
    # Model validation
    validation_model: Optional[str] = None        # trainer_name of deployed model to run on upload
    validation_labels: List[str] = []             # accepted prediction labels; empty = accept any
    validation_message: str = ""                  # rejection message shown to collector


class DatasetProfile(Document):
    org_id: str = ""
    name: str
    slug: Optional[str] = None                    # URL-friendly identifier, unique per org
    description: str = ""
    category: str = ""
    fields: List[DatasetField] = []
    status: str = "active"                        # draft | active | closed

    # Visibility & sharing
    visibility: str = "private"                   # private | public
    # Derived datasets — set when cloned or referenced from a public dataset
    source_dataset_id: Optional[str] = None       # original dataset this was cloned/referenced from
    reference_type: Optional[str] = None          # clone | reference
    # Cached stats for public listing (updated on entry submit)
    entry_count_cache: int = 0

    # Contributor task discovery — when True, appears in annotator task feed
    discoverable: bool = False
    # Optional allowlist of annotator emails; empty = open to all contributors
    contributor_allowlist: List[str] = []

    # Points / incentive system
    points_enabled: bool = False
    points_per_entry: int = 1                     # points awarded per field submission
    points_redemption_info: str = ""              # e.g. "100 pts = KES 10 airtime"

    # Location tracking
    require_location: bool = False                # if True, collector is prompted to share GPS location
    location_purpose: str = ""                    # shown to collector: why location is needed

    created_by: str = ""
    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)
    deleted_at: Optional[datetime] = None

    class Settings:
        name = "dataset_profiles"


class DatasetCollector(Document):
    org_id: str = ""
    dataset_id: str
    email: str
    name: str = ""
    token: str = Field(default_factory=lambda: secrets.token_urlsafe(32))
    status: str = "pending"                       # pending | active | completed
    invited_at: datetime = Field(default_factory=utc_now)
    last_active_at: Optional[datetime] = None
    entry_count: int = 0
    points_earned: int = 0
    deleted_at: Optional[datetime] = None

    class Settings:
        name = "dataset_collectors"


class EntryLocation(BaseModel):
    """GPS or IP-derived location captured at submission time."""
    lat: Optional[float] = None
    lng: Optional[float] = None
    accuracy: Optional[float] = None             # GPS accuracy in metres
    source: str = "ip"                           # "gps" | "ip"
    ip_address: Optional[str] = None
    country: Optional[str] = None               # ISO 3166-1 alpha-2 e.g. "KE"
    country_name: Optional[str] = None
    city: Optional[str] = None
    timezone: Optional[str] = None
    isp: Optional[str] = None


class DatasetEntry(Document):
    org_id: str = ""
    dataset_id: str
    collector_id: str
    field_id: str
    file_key: Optional[str] = None               # S3 object key
    file_mime: Optional[str] = None
    file_size_bytes: Optional[int] = None        # raw file size in bytes (set at upload time)
    text_value: Optional[str] = None             # text / number fields
    description: Optional[str] = None            # collector-provided description
    points_awarded: int = 0
    location: Optional[EntryLocation] = None     # GPS or IP-based location metadata
    captured_at: datetime = Field(default_factory=utc_now)

    class Settings:
        name = "dataset_entries"
